runtime/drivers/duckdb/connectors.go

```diff
@@ -8,34 +8,83 @@ import (
 	"strings"
 
 	"github.com/rilldata/rill/runtime/connectors"
+	"github.com/rilldata/rill/runtime/connectors/blob"
 	"github.com/rilldata/rill/runtime/connectors/localfile"
 	"github.com/rilldata/rill/runtime/drivers"
 	"github.com/rilldata/rill/runtime/pkg/fileutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
 )
 
+// increasing this limit can increase speed ingestion
+// but may increase bottleneck at duckdb or network/db IO
+// set without any benchamarks
+const concurrentBlobDownloadLimit = 32
+
 func (c *connection) Ingest(ctx context.Context, env *connectors.Env, source *connectors.Source) error {
 	err := source.Validate()
 	if err != nil {
 		return err
 	}
 
-	// Driver-specific overrides
-	// switch source.Connector {
-	// case "local_file":
-	// 	return c.ingestFile(ctx, env, source)
-	// }
-
+	// todo :: check if this exceptional handling can be merged
+	// locally uploaded file
 	if source.Connector == "local_file" {
-		return c.ingestFile(ctx, env, source)
+		conf, err := localfile.ParseConfig(source.Properties)
+		if err != nil {
+			return err
+		}
+		if !fileutil.HasMeta(conf.Path) {
+			return c.ingestFile(ctx, env, source)
+		}
 	}
 
-	path, err := connectors.ConsumeAsFile(ctx, env, source)
+	blobHandler, err := connectors.PrepareBlob(ctx, source)
 	if err != nil {
 		return err
 	}
-	defer os.Remove(path)
+	defer blobHandler.Close()
+	if len(blobHandler.FileNames) == 0 {
+		return fmt.Errorf("no filenames matching glob pattern")
+	}
+	c.logger.Info(fmt.Sprintf("matching files %v", blobHandler.FileNames))
+
+	// downloading first file and creating new table
+	if err := c.downloadAndIngest(ctx, source, blobHandler, blobHandler.FileNames[0], true); err != nil {
+		return err
+	}
+	// downloading other files in batch and appending to previoulsy created table
+	remainingFiles := blobHandler.FileNames[1:]
+	g, errCtx := errgroup.WithContext(ctx)
+	for i, file := range remainingFiles {
+		localFile := file
+		g.Go(func() error {
+			return c.downloadAndIngest(errCtx, source, blobHandler, localFile, false)
+		})
+		if (i+1)%concurrentBlobDownloadLimit == 0 {
+			if err := g.Wait(); err != nil {
+				return err
+			}
+		}
+	}
+	return g.Wait()
```

Comment:
Is calling `Wait` many times on the same `errgroup` documented behaviour? Also, this will wait for all downloads in a batch to complete before starting the next batch. Instead, you could consider:
1. Starting `min(len(FileNames), concurrentBlobDownloadLimit)` goroutines
2. You can track the `idx` of the next file to process using an `atomic.Int` (for concurrency safety)
3. In each goroutine, have a loop that gets files as long as all have not been taken (e.g. `for i := idx.Add(1); i < len(FileNames)`)